translate: State as a snapshot (#468)
reactjs/tr.react.dev: src/content/learn/state-as-a-snapshot.md

Bir fotoğraf veya film karesinin aksine, döndürdüğünüz kullanıcı arayüzü "anlık görüntü" etkileşimlidir. Bu, girdilere yanıt olarak neyin gerçekleşeceğini belirten olay yöneticileri gibi mantık içerir. React, ekranı bu anlık görüntüyle eşleştirmek ve olay yöneticilerini bağlamak için güncelleme yapar. Sonuç olarak, bir butona basmak JSX'inizdeki tıklama yöneticisini tetikleyecektir.

React bir bileşeni yeniden render ettiğinde: 

1. React fonksiyonunuzu yeniden çağırır.
2. Fonksiyonunuz yeni bir JSX anlık görüntüsü döner.
3. React ardından, döndürdüğünüz anlık görüntüye göre ekranı günceller

<IllustrationBlock sequential>
    <Illustration caption="React fonksiyonu çalıştırıyor" src="/images/docs/illustrations/i_render1.png" />
    <Illustration caption="Anlık görüntüyü hesaplıyor" src="/images/docs/illustrations/i_render2.png" />
    <Illustration caption="DOM ağacını güncelliyor" src="/images/docs/illustrations/i_render3.png" />
</IllustrationBlock>

> Bir bileşenin hafızası olarak, state fonksiyonunuz döndüğünde ortadan kaybolan sıradan bir değişken değildir. State aslında fonksiyonunuzun dışında React’in içinde yaşar sanki bir rafta duruyormuş gibi. React bileşeninizi çağırdığında o belirli render için durumun bir anlık görüntüsünü size verir. Bileşeniniz JSX içinde hesaplanmış taze bir takım prop ve olay yöneticileriyle birlikte **o render için state değerlerini kullanarak** kullanıcı arayüzünün bir anlık görüntüsünü döndürür.

Reviewer: ```suggestion
Bir bileşenin hafızası olarak state, fonksiyonunuzun yürütülmesi sona erdiğinde ortadan kaybolan sıradan bir değişken değildir. State sanki bir rafta duruyormuşçasına fonksiyonunuzun dışında  React’in kendi içinde yaşar. React bileşeninizi çağırdığında, söz konusu render için state'in anlık görüntüsünü verir. Bileşeniniz, **ilgili render'daki state değeri kullanılarak** hesaplanmış JSX'indeki yeni prop'lar ve olay yöneticileriyle birlikte kullanıcı arayüzünüzün anlık görüntüsünü döndürür.
```
Bileşeniniz, JSX'inde yeni bir dizi sahne ve olay işleyicisi ile kullanıcı arayüzünün bir anlık görüntüsünü döndürür ve bunların tümü **o render'daki durum değerlerini kullanarak hesaplanır!

<IllustrationBlock sequential>

>   <Illustration caption="React'a state'i güncellemesini söylersiniz." src="/images/docs/illustrations/i_state-snapshot1.png" />
>   <Illustration caption="React, state değerini günceller." src="/images/docs/illustrations/i_state-snapshot2.png" />
>   <Illustration caption="React, bileşene state değerinin bir anlık görüntüsünü aktarır." src="/images/docs/illustrations/i_state-snapshot3.png" />

Reviewer: ```suggestion
  <Illustration caption="React'e state'i güncellemesini söylersiniz" src="/images/docs/illustrations/i_state-snapshot1.png" />
  <Illustration caption="React state değerini günceller" src="/images/docs/illustrations/i_state-snapshot2.png" />
  <Illustration caption="React bileşene state değerinin anlık görüntüsünü aktarır" src="/images/docs/illustrations/i_state-snapshot3.png" />
```

</IllustrationBlock>

> İşleyişin nasıl çalıştığını göstermek için küçük bir deney yapalım. Bu örnekte, "+3" butonuna tıkladığınızda `setNumber(number + 1)` üç kez çağrıldığı için sayaçın üç kez artmasını bekleyebilirsiniz. 

Reviewer: ```suggestion
İşte size bunun nasıl çalıştığını gösteren küçük bir deney. Bu örnekte, "+3" butonuna tıkladığınızda `setNumber(number + 1)` üç kez çağrıldığı için sayacın üç kez artmasını bekleyebilirsiniz. 
```

> “+3” butonuna tıkladığınızda ne olduğunu görelim

Reviewer: ```suggestion
"+3" butonuna tıkladığınızda ne olduğunu görün:
```

</Sandpack>

`number` değerinin her tıklamada bir kez arttığına dikkat edin!

> **State’i ayarlamak sadece *sonraki* render işlemi için değiştirir.** İlk render sırasında, `number` değeri `0` idi. Bu yüzden, o render'ın `onClick` işleyicisinde `setNumber(number + 1)` çağrıldıktan sonra bile, `number` değeri hala `0` olarak kalır.

Reviewer: ```suggestion
**State’i ayarlamak yalnızca *sonraki* render işlemi için değerini değiştirir.** İlk render sırasında, `number` değeri `0` idi. Bu yüzden, *o render'ın* `onClick` işleyicisinde `setNumber(number + 1)` çağrıldıktan sonra bile `number` değeri değişmez ve `0` olarak kalır.
```

}}>+3</button>
```

> İşte bu butonun tıklama yöneticisine React'e ne yapması gerektiğini söylediği:

Reviewer: ```suggestion
Bu butonun tıklama işleyicisinin React'e yapmasını söylediği şey aşağıdaki gibidir:
```

}}>+3</button>
```

Bu nedenle, düğmeye tekrar tıkladığınızda sayaç önce `2` olarak ayarlanır, ardından bir sonraki tıklamada `3` olarak ayarlanır ve böyle devam eder.

## Zaman içerisinde state {/*state-over-time*/}

> Bu eğlenceliydi. Bu düğmeye tıklamanın ne uyarı vereceğini tahmin etmeye çalışın:

Reviewer: ```suggestion
Pekala, bu eğlenceliydi. Bu düğmeye tıklamanın ne uyarı vereceğini tahmin etmeye çalışın:
```

</Sandpack>

Önceden belirttiğimiz yerine koyma yöntemini kullanarak, butona tıklandığında ekranda "0" şeklinde bir uyarı mesajı görüntülenmesini tahmin edebilirsiniz.

```js
setNumber(0 + 5);
alert(0);
```

> Eğer uyarıya bir zamanlayıcı ekler ve bu zamanlayıcı yalnızca bileşen yeniden render edildikten _sonra_ tetiklenirse ne olur ? “0” mı yoksa “5” mi der? Bir tahminde bulunun!

Reviewer: ```suggestion
Peki ya uyarıya bir zamanlayıcı koyarsanız ve yalnızca bileşen yeniden render edildikten _sonra_ tetiklenirse ne olur? “0” mı yoksa “5” mi der? Bir tahminde bulunun!
```

}, 3000);
```

React'te depolanan state, uyarı çalıştığında değişmiş olabilir, ancak kullanıcının etkileşimde bulunduğu zamandaki durumun bir anlık görüntüsü kullanılarak zamanlanmıştır.

> *Bir state değişkeninin değeri bir render işlemi içinde asla değişmez,** hatta yöneticisinin kodu asenkron olsa bile. İçindeki *o render’ın*  onClick yöneticisinde, `setNumber(number + 5)` çağrılmış olsa bile, `number`'ın değeri hala `0` olarak kalır. `Number`'ın değeri React bileşeninizi çağırarak kullanıcı arayüzü'nün "anlık görüntüsü'nü" alırken "sabitlenmiştir".

Reviewer: ```suggestion
**Bir state değişkeninin değeri render işlemi içinde asla değişmez,** hatta olay yöneticisinin kodu asenkron olsa bile. *O render’ın*  onClick yöneticisinde, `setNumber(number + 5)` çağrılmış olsa bile `number`'ın değeri `0` olarak kalır. State değeri React bileşeninizi çağırdığında kullanıcı arayüzünün "anlık görüntüsünü" alırken "sabitlenmiştir".
```

İşte bu durumun olay yöneticilerinizi zamanlama hatalarına karşı daha yatkın hale getiren bir örneği. Aşağıda, beş saniyelik bir gecikmeyle mesaj gönderen bir form bulunmaktadır. Bu senaryoyu hayal edin:

> 1. “Send” butonuna basarak “Hello” mesajını Alice’e gönderiyorsunuz

Reviewer: ```suggestion
1. “Send” butonuna basarak “Merhaba” mesajını Alice’e gönderiyorsunuz
```

> 2. Beş saniyelik gecikme bitmeden önce, “To” alanının değerini “Bob” olarak değiştiriyorsunuz.

Reviewer: ```suggestion
2. Beş saniyelik gecikme bitmeden “Kime” alanının değerini “Bob” olarak değiştiriyorsunuz.
```

> `alert`'in ne görüntüleyeceğini bekliyorsunuz? "You said Hello to Alice" mı yoksa "You said Hello to Bob" mı? Bildiklerinizden yola çıkarak bir tahminde bulunun ve deneyin:

Reviewer: ```suggestion
`alert`'in ne göstermesini beklersiniz? "Alice'e merhaba dedin." mi yoksa "Bob'a merhaba dedin." mi? Bildiklerinizden yola çıkarak bir tahminde bulunun ve deneyin:
```

Reviewer: Altındaki örnekte bulunan metinleri de Türkçe halleriyle değiştirebilirsek süper olur.

</Sandpack>

> **React, bir render içindeki olay yöneticilerindeki state değerlerini “sabit” tutar.** Kod çalışırken durumun değişip değişmediği konusunda endişelenmenize gerek yoktur.

Reviewer: ```suggestion
**React, state değerlerini render'ın olay yöneticileri içinde “sabit” tutar.** Kod çalışırken durumun değişip değişmediği konusunda endişelenmeniz gerekmez.
```

> Ancak bir yeniden render alma öncesi en son state’i okumak isterseniz ne yapabilirsiniz? Bir sonraki sayfada ele alınan [state güncelleme fonksiyonu](/learn/queueing-a-series-of-state-updates)'nu kullanmak isteyeceksiniz!

Reviewer: ```suggestion
Peki ya state'in en son değerini yeniden render tetiklenmeden önce okumak isterseniz? Bir sonraki sayfada ele alınan [state güncelleme fonksiyonu](/learn/queueing-a-series-of-state-updates)nu kullanmak isteyeceksiniz!
```

<Recap>

> * State’i ayarlanması yeni bir render ister.
> * React State’i bileşeninizin dışında, sanki bir raf üzerinde depolar.
> * `useState` hook'unu çağırdığınızda, React *o render* için state'in bir anlık görüntüsünü verir.
> * Değişkenler ve olay yöneticileri yeniden render alınmasıyla “kurtulamazlar”. Her render’ın kendi olayı yöneticileri vardır.
> * Her render (ve içindeki fonksiyonlar), her zaman React’in *o* render’a verdiği state’in anlık görüntüsünü görecektir.
> * Renderlanmış JSX hakkında ne düşündüğünüze benzer bir şekilde, zihinsel olarak olay yöneticileri içinde state’i yerine koyabilirsiniz, 
> * Geçmişte oluşturulmuş olay yöneticileri, oluşturuldukları render’ın state değerine sahiptirler.

Reviewer: ```suggestion
* State’in ayarlanması yeni bir render isteği yapar.
* React state’i bileşeninizin dışında, sanki bir raftaymış gibi, depolar.
* `useState`'i çağırdığınızda, React *o render* için state'in anlık görüntüsünü verir.
* Değişkenler ve olay yöneticileri yeniden render işlemi esnasında “hayatta kalmazlar”. Her render’ın kendi olayı yöneticileri vardır.
* Tüm render'lar (ve içindeki fonksiyonlar), daima React’in *o* render’a verdiği state anlık görüntüsünü görür.
* Renderlanmış JSX hakkında düşündüğünüze benzer şekilde, zihinsel olarak olay yöneticileri için de state’i yerine koyabilirsiniz.
* Geçmişte oluşturulmuş olay yöneticileri, oluşturuldukları render'ın state değerine sahiptir.
```

</Recap>

<Challenges>

#### Bir trafik ışığı uygulayın {/*implement-a-traffic-light*/}

> İşte butona basıldığında yaya geçidi ışığını açıp kapatan bir bileşen:

Reviewer: ```suggestion
İşte butona basıldığında geçiş yapan bir yaya geçidi ışık bileşeni:
```

Reviewer: Altındaki örnekte bulunan metinleri Türkçe halleriyle değiştirirsek süper olur.

</Sandpack>

> İlgili tıklama yöneticisine bir `Alarm` ekleyelim. Yeşil ışık yanıyorsa ve `Walk` yazıyorsa, butona tıklanıldığında `Walk is Next` demesi gerekiyor. Kırmızı ışık yanıyorsa ve “Dur” yazıyorsa, butona tıklandığında `Walk is Next` demesi gerekiyor.

Reviewer: ```suggestion
Tıklama yöneticisine bir `alert` ekleyin. Yeşil ışık yanıyor ve `Yürü` yazıyorsa, butona tıklamak `Sıradaki: Dur` demelidir. Kırmızı ışık yanıyor ve “Dur” yazıyorsa, butona tıklamak `Sıradaki: Yürü` demelidir.
```

Reviewer: Makale boyunca kullanılan `alert` JavaScript'teki `alert` metoduna karşılık geliyor. Çevirilmemeli.

> `Alarm`'ı `setWalk` çağrısının öncesine veya sonrasına koymak arasında bir fark oluşturur mu?

Reviewer: ```suggestion
`alert`'i `setWalk` çağrısından önce veya sonra koymanız bir fark yaratır mı?
```

<Solution>

> `Alarmınız` şu şekilde görünmeli:

Reviewer: ```suggestion
`alert`ünüz şu şekilde görünmelidir:
```

</Sandpack>

> `setWalk` çağrısının öncesinde veya sonrasında olması fark etmez. walk'un render işlemi sırasındaki değeri sabittir. setWalk çağrısı, sadece *sonraki* render işlemi için değeri değiştirir, ancak önceki render işlemi için olay yöneticisini etkilemez.

Reviewer: ```suggestion
`setWalk` çağrısının öncesinde veya sonrasında olması fark etmez. `walk`'un render sırasındaki değeri sabittir. `setWalk` çağrısı, yalnızca *sonraki* render için değeri değiştirir ancak önceki render'daki olay yöneticisini etkilemez.
```

Bu satır ilk başta mantıksız görünebilir:

```js

> alert(walk ? 'Stop is next' : 'Walk is next');

Reviewer: ```suggestion
alert(walk ? 'Sıradaki: Dur' : 'Sıradaki: Yürü');
```

```

> Ancak, "Eğer trafik lambası 'Walk now' gösteriyorsa, mesaj 'Stop is next' demelidir." şeklinde okursanız, mantıklı olur. Olay yöneticinizdeki walk değişkeni, ilgili render'in walk değeriyle eşleşir ve değişmez.

Reviewer: ```suggestion
Ancak şu şekilde okursanız mantıklı olur: "Trafik ışığı 'Yürü' gösteriyorsa, mesaj 'Sıradaki: Dur' demelidir." Olay işleyicinizdeki `walk` değişkeni, render'ın `walk` değeriyle eşleşir ve değişmez.
```

> Bu doğru olduğunu doğrulamak için yerine koyma yöntemini uygulayarak kontrol edebilirsiniz. walk değeri true olduğunda şunu elde edersiniz:

Reviewer: ```suggestion
Bunun doğruluğunu yerini koyma yöntemini uygulayarak kontrol edebilirsiniz. 'walk''ın değeri 'true' olduğunda şunu elde edersiniz:
```

</h1>
```

> "Change to Stop" butonuna tıklamak, walk değeri false olarak ayarlanmış bir render'ı sıraya alır ve "Stop is next" uyarısını verir.

Reviewer: ```suggestion
Yani "Dur Olarak Değiştir" butonuna tıklamak, `walk`'ın `false` olarak ayarlandığı render işlemini kuyruğa alır ve "Sıradaki: Dur" uyarısını verir.
```
